### PR TITLE
[refactor]: Refactor Frontend Trace OpenTelemetry Implementation

### PR DESCRIPTION
**Pull request by oandreeva-nv**
#### What does the PR do?

This PR introduces an improvement to our frontend OpenTelemetry tracing implementation. The main goal of this refactoring is to enhance the reliability and flexibility of the system.

Key Changes:
-  Introduced a Stack-Based Approach
    - The previous implementation  relied on specific span names to preserve the trace hierarchy. 
    - Current solution uses stacks and doesn't need to know what type of span was started.
    - For ensemble and bls models, any number of sub-traces can be spawned. Since we preserve `parent_id` for every trace, I utilize this and introduced an unordered map to keep spans in stacks. Each trace and sub-trace has its own stack. When stack is empty, I use `parent_id` to check the top of the parent's stack to get parent span's information.
    - This refactor also helps the following `Custom backend tracing` implementation to avoid handling of the complex logic for multiple nested custom spans, i.e. we simply look at the top of the stack.

#### Checklist
- [X] PR title reflects the change and is of format `<commit_type>: <Title>`
- [X] Changes are described in the pull request.
- [ ] Related issues are referenced.
- [X] Populated [github labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) field
- [ ] Added [test plan](#test-plan) and verified test passes.
- [] Verified that the PR passes existing CI.
- [X] Verified copyright is correct on all changed files.
- [ ] Added _succinct_ git squash message before merging [ref](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html).
- [ ] All template sections are filled out.
- [ ] Optional: Additional screenshots for behavior/output changes with before/after.

#### Commit Type:
Check the [conventional commit type](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type)
box here and add the label to the github PR.
- [ ] build
- [ ] ci
- [ ] docs
- [ ] feat
- [ ] fix
- [ ] perf
- [X] refactor
- [ ] revert
- [ ] style
- [ ] test

#### Related PRs:

N/A 

#### Where should the reviewer start?

trace.h 

#### Test plan:

- CI Pipeline ID: 16139557

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)
N/A

### CODE REVIEW

File: src/tracer.cc [hunk 1]
```diff
@@ -501,15 +512,15 @@ TraceManager::ProcessOpenTelemetryParameters(
...
+    TRITONSERVER_InferenceTrace* trace,
     TRITONSERVER_InferenceTraceActivity activity, uint64_t timestamp_ns,
     uint64_t trace_id)
 {
   uint64_t parent_id;
   LOG_TRITONSERVER_ERROR(
       TRITONSERVER_InferenceTraceParentId(trace, &parent_id),
       "getting trace parent id");
-  std::string parent_span_key = "";
+  auto span_parent_id = parent_id;
```
Comment (GuanLuo) [inline src/tracer.cc:523]: Just use parent_id and zero-initialize it first?

File: src/tracer.cc [hunk 2]
```diff
@@ -580,45 +583,37 @@ TraceManager::Trace::StartSpan(
...
-  // the context and provide it through StartSpanOptions to the child span
-  if (!parent_span_key.empty() && otel_context_.HasKey(parent_span_key)) {
-    auto parent_span = opentelemetry::nostd::get<
-        opentelemetry::nostd::shared_ptr<otel_trace_api::Span>>(
-        otel_context_.GetValue(parent_span_key));
-    options.parent = parent_span->GetContext();
+  // If the new span is a child span, we need to retrieve its parent and
+  // provide it through StartSpanOptions to the child span
+  if (span_stacks_.find(trace_id) != span_stacks_.end() &&
+      !span_stacks_[trace_id]->empty()) {
```
Comment (GuanLuo) [inline src/tracer.cc:589]: Won't this be entered twice if the caller: start span -> end span -> start span? Will this mistakenly treat the trace to "have" a parent?

**Review by Tabrizian — COMMENTED** [review]
Could you add an example for how this looks like for t

File: src/tracer.cc (hunk 2 above)
Comment (oandreeva-nv) [reply src/tracer.cc:589]: Every `end span` event pops span from the stack, so it doesn't exist anymore. 
The trace frontend currently controls root span creation, request spans and compute spans. The passed `trace_id` in this method obides the logic, explained here: https://github.com/triton-inference-server/server/pull/7390/files#diff-c05667609085c993fd282ff18d0429945c44492b95c0d01fcbeec6946fda2db2R525-R534

File: src/tracer.h [hunk 3]
```diff
@@ -315,58 +323,44 @@ class TraceManager {
...
-
-    /// Adds event to the span, which is retrieved from OpenTelemetry context
-    /// with the provided `span_key`. If activity is
-    /// TRITONSERVER_TRACE_REQUEST_START, or TRITONSERVER_TRACE_COMPUTE_START,
-    /// starts a new span and adds it to `otel_context_`.
+    void EndSpan(const uint64_t& raw_timestamp_ns, uint64_t trace_id);
+
+    /// Adds event to the span on the top of the stack, related to trace
+    /// with `trace_id`. If activity is TRITONSERVER_TRACE_REQUEST_START,
+    /// or TRITONSERVER_TRACE_COMPUTE_START, starts a new span and adds it
```
[the remark this answers is not in the document]
Comment (oandreeva-nv) [reply src/tracer.h:345]: Could you please clarify?

**Comment by oandreeva-nv** [comment]
@Tabrizian this is just a re-factor for the frontend. The actual custom backend implementation PR will be up shortly

File: src/tracer.cc (hunk 1 above)
Comment (oandreeva-nv) [reply src/tracer.cc:523]: I need both, since later on L563, I need true `parent_id`

**Review by GuanLuo — APPROVED** [review]

**Pull request merged**